src/Microsoft.AspNet.OData.Shared/Query/SkipTokenQueryOption.cs

```diff
@@ -0,0 +1,183 @@
+        /// <param name="queryOptionParser">The <see cref="ODataQueryOptionParser"/> which is used to parse the query option.</param>
+        public SkipTokenQueryOption(string rawValue, ODataQueryContext context, ODataQueryOptionParser queryOptionParser)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull("context");
+            }
+
+            if (String.IsNullOrEmpty(rawValue))
+            {
+                throw Error.ArgumentNullOrEmpty("rawValue");
+            }
+
+            if (queryOptionParser == null)
+            {
+                throw Error.ArgumentNull("queryOptionParser");
+            }
+
+            RawValue = rawValue;
+            Validator = SkipTokenQueryValidator.GetSkipTokenQueryValidator(context);
+            SkipTokenHandler = GetSkipTokenHandler(context);
+            Context = context;
+            _queryOptionParser = queryOptionParser;
+        }
+
+        // This constructor is intended for unit testing only.
+        internal SkipTokenQueryOption(string rawValue, ODataQueryContext context)
+        {
+            if (context == null)
+            {
+                throw Error.ArgumentNull("context");
+            }
+
+            if (String.IsNullOrEmpty(rawValue))
+            {
+                throw Error.ArgumentNullOrEmpty("rawValue");
+            }
+
+            RawValue = rawValue;
+            Validator = SkipTokenQueryValidator.GetSkipTokenQueryValidator(context);
+            _queryOptionParser = new ODataQueryOptionParser(
+                context.Model,
+                context.ElementType,
+                context.NavigationSource,
+                new Dictionary<string, string> { { "$skiptoken", rawValue } },
+                context.RequestContainer);
+        }
+
+        /// <summary>
+        /// Gets the raw $skiptoken value.
+        /// </summary>
+        public string RawValue { get; private set; }
+
+        /// <summary>
+        /// Gets and sets the given <see cref="ODataQueryContext"/>.
+        /// </summary>
+        public ODataQueryContext Context { get; set; }
+
+        /// <summary>
+        /// Generates the nextlink value and consumes the skiptoken value.
+        /// </summary>
+        private SkipTokenHandler SkipTokenHandler { get; set; }
+
+        /// <summary>
+        /// Gets the value of the $skiptoken.
+        /// </summary>
+        public string Value
+        {
+            get
+            {
+                if (_value == null)
+                {
+                    string skipValue = _queryOptionParser.ParseSkipToken();
+                    _value = skipValue;
+                }
+                return _value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the SkipToken Query Validator.
+        /// </summary>
+        public SkipTokenQueryValidator Validator { get; set; }
```

Comment:
> set; } [](start = 55, length = 7)

Make it as public set is not a good design. Maybe file an issue to track all query option classes.

Reply:
[https://github.com/OData/WebApi/issues/1805](https://github.com/OData/WebApi/issues/1805)

---
In reply to: [281842108](https://github.com/OData/WebApi/pull/1724#discussion_r281842108) [](ancestors = 281842108)

Reply:
Can we make this private for now, so that if we are able to do cleanup as per 1805, we don't have a breaking change (at least for skiptokenvalidator)?

---
In reply to: [282645426](https://github.com/OData/WebApi/pull/1724#discussion_r282645426) [](ancestors = 282645426,281842108)